[pytorch] Make some singletons leaky, to deal with module destruct order.

jjlilley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30103 [pytorch] Make some singletons leaky, to deal with module destruct order.**

RRefContext is a conventional singleton, used by rref.cpp. At module teardown
time, it's not defined whether rref_context.cpp or rref.cpp will be destroyed first.

We were observing a SIGSEGV because RRefContext is destroyed before a dangling
~UserRRef() call is able to execute. Particularly, the underlying
ctx.agent()->getWorkerInfo(ownerId_) call failed.

This change just avoids the SIGSEGV by forcing an intentional leak, though we still
need to deal with why there's a dangling UserRref at module destruction time.

Additionally, we add this pattern to DistAutogradContainer, DistEngine, and
PythonRpcHandler's singletons.

Differential Revision: [D18598208](https://our.internmc.facebook.com/intern/diff/D18598208/)

Commit:
[pytorch] Make RRefContext a leaky singleton, to deal with module destruct order.

RRefContext is a conventional singleton, used by rref.cpp. At module teardown
time, it's not defined whether rref_context.cpp or rref.cpp will be destroyed first.

We were observing a SIGSEGV because RRefContext is destroyed before a dangling
~UserRRef() call is able to execute. Particularly, the underlying
ctx.agent()->getWorkerInfo(ownerId_) call failed.

This change just avoids the SIGSEGV by forcing an intentional leak, though we still
need to deal with why there's a dangling UserRref at module destruction time.

Differential Revision: [D18598208](https://our.internmc.facebook.com/intern/diff/D18598208/)

[ghstack-poisoned]

Commit:
Update on "[pytorch] Make RRefContext a leaky singleton, to deal with module destruct order."


RRefContext is a conventional singleton, used by rref.cpp. At module teardown
time, it's not defined whether rref_context.cpp or rref.cpp will be destroyed first.

We were observing a SIGSEGV because RRefContext is destroyed before a dangling
~UserRRef() call is able to execute. Particularly, the underlying
ctx.agent()->getWorkerInfo(ownerId_) call failed.

This change just avoids the SIGSEGV by forcing an intentional leak, though we still
need to deal with why there's a dangling UserRref at module destruction time.

Differential Revision: [D18598208](https://our.internmc.facebook.com/intern/diff/D18598208/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -8,8 +8,10 @@ namespace distributed {
-  static RRefContext context(RpcAgent::getDefaultRpcAgent());
-  return context;
+  // Intentionally leaky singleton; this module may be destructed before
+  // rref.cpp, which may use getInstance() in its module destructor.
```
Just to make sure I understand this. `RRefContext` holds a `shared_ptr` to the `RpcAgent`, so the agent will also be alive when `~UserRRef` access it right?

jjlilley replied:
Totally agree that we need more changes.

There are two problems here:
 1. the python code isn't shutting down rrefs before the program is about to destruct.
 2. when (1) happens, the program segfaults

This particularly change is all about a 2-line workaround to address (2), rather than the real fix for (1).

mrshenli reviewed (commented):
We might need more changes. When calling `join_rpc()`, `ProcessGroupAgent` would send out shutdown messages and then join the listener thread. So it means that the message sent in `~UserRRef` might not be processed on the owner side, which is likely to result in a timeout in the thread pool sending thread.

mrshenli commented:
created #30105 to track nuking UserRRefs

pritamdamania87 commented:
@jjlilley Thanks for looking into this! Would it be possible to fix the other singletons in this PR too? We have singletons for DistAutogradContainer, DistEngine and PythonRpcHandler (basically a search for getInstance() should uncover all singletons) as well. Also, would be nice to reference https://github.com/pytorch/pytorch/issues/27412 in this PR since we're tracking it there.

Commit:
Update on "[pytorch] Make some singletons leaky, to deal with module destruct order."


RRefContext is a conventional singleton, used by rref.cpp. At module teardown
time, it's not defined whether rref_context.cpp or rref.cpp will be destroyed first.

We were observing a SIGSEGV because RRefContext is destroyed before a dangling
~UserRRef() call is able to execute. Particularly, the underlying
ctx.agent()->getWorkerInfo(ownerId_) call failed.

This change just avoids the SIGSEGV by forcing an intentional leak, though we still
need to deal with why there's a dangling UserRref at module destruction time.

Additionally, we add this pattern to DistAutogradContainer, DistEngine, and
PythonRpcHandler's singletons.

Differential Revision: [D18598208](https://our.internmc.facebook.com/intern/diff/D18598208/)

[ghstack-poisoned]

jjlilley commented:
> @jjlilley Thanks for looking into this! Would it be possible to fix the other singletons in this PR too? We have singletons for DistAutogradContainer, DistEngine and PythonRpcHandler (basically a search for getInstance() should uncover all singletons) as well. Also, would be nice to reference #27412 in this PR since we're tracking it there.

Sure, I just went ahead and uploaded a version that uses that pattern in those other singletons as well.

zhaojuanmao reviewed (approved):
looks good, please land after tests are green, thanks!

Commit:
Update on "[pytorch] Make some singletons leaky, to deal with module destruct order."


RRefContext is a conventional singleton, used by rref.cpp. At module teardown
time, it's not defined whether rref_context.cpp or rref.cpp will be destroyed first.

We were observing a SIGSEGV because RRefContext is destroyed before a dangling
~UserRRef() call is able to execute. Particularly, the underlying
ctx.agent()->getWorkerInfo(ownerId_) call failed.

This change just avoids the SIGSEGV by forcing an intentional leak, though we still
need to deal with why there's a dangling UserRref at module destruction time.

Additionally, we add this pattern to DistAutogradContainer, DistEngine, and
PythonRpcHandler's singletons.

Differential Revision: [D18598208](https://our.internmc.facebook.com/intern/diff/D18598208/)

[ghstack-poisoned]

Commit:
Update on "[pytorch] Make some singletons leaky, to deal with module destruct order."


RRefContext is a conventional singleton, used by rref.cpp. At module teardown
time, it's not defined whether rref_context.cpp or rref.cpp will be destroyed first.

We were observing a SIGSEGV because RRefContext is destroyed before a dangling
~UserRRef() call is able to execute. Particularly, the underlying
ctx.agent()->getWorkerInfo(ownerId_) call failed.

This change just avoids the SIGSEGV by forcing an intentional leak, though we still
need to deal with why there's a dangling UserRref at module destruction time.

Additionally, we add this pattern to DistAutogradContainer, DistEngine, and
PythonRpcHandler's singletons.

Differential Revision: [D18598208](https://our.internmc.facebook.com/intern/diff/D18598208/)

[ghstack-poisoned]

jjlilley commented:
I'm planning to merge https://github.com/pytorch/pytorch/pull/30172 instead, which only addresses RRefContext, rather than the other 3 singletons as well.

It looks like there are some side-effects from the other singletons not hitting module destructors, and I don't want to unnecessarily destabilize the branch.
